tools/shared/src/main/scala/org/scalajs/core/tools/linker/frontend/optimizer/OptimizerCore.scala

```diff
@@ -3302,9 +3327,23 @@ private[optimizer] abstract class OptimizerCore(
     } else {
       val refinedType = value.tpe
       value match {
-        case PreTransBlock(stats, result) =>
+        case PreTransBlock(bindingsAndStats, result) =>
           withNewLocalDef(binding.copy(value = result))(buildInner) { tresult =>
-            cont(PreTransBlock(stats, tresult))
+            val preTransBlock = tresult match {
+              case tresult: PreTransBlock =>
+                PreTransBlock(bindingsAndStats, tresult)
+              case tresult: PreTransLocalDef =>
+                PreTransBlock(bindingsAndStats, tresult)
+              case PreTransRecordTree(tree, tpe, cancelFun) =>
+                PreTransRecordTree(
+                    finishTransformBindings(bindingsAndStats, tree),
+                    tpe, cancelFun)
+              case PreTransTree(tree, tpe) =>
+                PreTransTree(
+                    finishTransformBindings(bindingsAndStats, tree),
+                    tpe)
+            }
```

Comment:
They are not actually factorable. One works with an only `PreTransBinding` and the other with a `List[BindingOrStat]`. However I have extracted those functionalities in two helper methods defined next to each other, and with more doc, so that it's obvious they do "the same thing" but not on the same data types.